File: project/backend/test_backend_api.mjs
import fetch from "node-fetch";

const BASE_URL = "http://localhost:8000/api";

// ANSI escape codes for colors
const GREEN = "\x1b[32m";
const RED = "\x1b[31m";
const BLUE = "\x1b[34m";
const RESET = "\x1b[0m";

function printResult(testCase, description, result, expected) {
  if (result === expected) {
    console.log(`${GREEN}[PASS]${RESET} ${testCase}: ${description}`);
  } else {
    console.log(`${RED}[FAIL]${RESET} ${testCase}: ${description}`);
    console.log(`  Expected: ${expected}, Got: ${result}`);
  }
}

async function testSignup() {
  const url = `${BASE_URL}/auth/signup`;
  const payload = {
    fname: "John",
    lname: "Doe",
    email: "john.doe@example.com",
    username: "johndoe",
    password: "Password123",
  };

  try {
    const response = await fetch(url, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(payload),
    });

    printResult("TC-BACK-001", "Verify user signup", response.status, 201);
  } catch (error) {
    console.error(`${RED}[ERROR]${RESET} ${error.message}`);
  }
}

async function testLogin() {
  const url = `${BASE_URL}/auth/login`;

  // Valid credentials
  const validPayload = { username: "johndoe", password: "Password123" };
  try {
    const validResponse = await fetch(url, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(validPayload),
    });
    printResult("TC-BACK-002", "Login with valid credentials", validResponse.status, 200);
  } catch (error) {
    console.error(`${RED}[ERROR]${RESET} ${error.message}`);
  }

  // Invalid password
  const invalidPasswordPayload = { username: "johndoe", password: "WrongPassword" };
  try {
    const invalidPasswordResponse = await fetch(url, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(invalidPasswordPayload),
    });
    printResult("TC-BACK-002", "Login with invalid password", invalidPasswordResponse.status, 401);
  } catch (error) {
    console.error(`${RED}[ERROR]${RESET} ${error.message}`);
  }

  // User not found
  const invalidUserPayload = { username: "nonexistentuser", password: "Password123" };
  try {
    const invalidUserResponse = await fetch(url, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(invalidUserPayload),
    });
    printResult("TC-BACK-002", "Login with nonexistent user", invalidUserResponse.status, 404);
  } catch (error) {
    console.error(`${RED}[ERROR]${RESET} ${error.message}`);
  }
}

async function testAdminPermissions(userId) {
  const url = `${BASE_URL}/admin/permissions/${userId}`;

  // Grant admin privileges
  try {
    const grantResponse = await fetch(url, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
    });
    printResult("TC-BACK-004", "Grant admin privileges", grantResponse.status, 200);
  } catch (error) {
    console.error(`${RED}[ERROR]${RESET} ${error.message}`);
  }

  // Remove admin privileges
  try {
    const removeResponse = await fetch(url, {
      method: "DELETE",
      headers: { "Content-Type": "application/json" },
    });
    printResult("TC-BACK-004", "Remove admin privileges", removeResponse.status, 200);
  } catch (error) {
    console.error(`${RED}[ERROR]${RESET} ${error.message}`);
  }
}

async function testGetAllUsers() {
  const url = `${BASE_URL}/admin/users`;

  try {
    const response = await fetch(url, {
      method: "GET",
      headers: { "Content-Type": "application/json" },
    });

    printResult("TC-BACK-005", "Retrieve all user accounts", response.status, 200);
  } catch (error) {
    console.error(`${RED}[ERROR]${RESET} ${error.message}`);
  }
}

async function testUserManagement(username) {
  const url = `${BASE_URL}/admin/users/${username}`;

  // Find user
  try {
    const findResponse = await fetch(url, {
      method: "GET",
      headers: { "Content-Type": "application/json" },
    });
    printResult("TC-BACK-006", "Find user by username", findResponse.status, 200);
  } catch (error) {
    console.error(`${RED}[ERROR]${RESET} ${error.message}`);
  }

  // Remove user
  try {
    const removeResponse = await fetch(url, {
      method: "DELETE",
      headers: { "Content-Type": "application/json" },
    });
    printResult("TC-BACK-006", "Remove user by username", removeResponse.status, 200);
  } catch (error) {
    console.error(`${RED}[ERROR]${RESET} ${error.message}`);
  }
}

async function testUserMe() {
  const url = `${BASE_URL}/user/me`;

  // Retrieve user info
  try {
    const retrieveResponse = await fetch(url, {
      method: "GET",
      headers: { "Content-Type": "application/json" },
    });
    printResult("TC-BACK-007", "Retrieve user info", retrieveResponse.status, 200);
  } catch (error) {
    console.error(`${RED}[ERROR]${RESET} ${error.message}`);
  }

  // Update user info
  const updatePayload = {
    fname: "Jane",
    lname: "Doe",
    email: "jane.doe@example.com",
    username: "janedoe",
    password: "NewPassword123",
  };
  try {
    const updateResponse = await fetch(url, {
      method: "PATCH",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(updatePayload),
    });
    printResult("TC-BACK-007", "Update user info", updateResponse.status, 200);
  } catch (error) {
    console.error(`${RED}[ERROR]${RESET} ${error.message}`);
  }

  // Delete user account
  try {
    const deleteResponse = await fetch(url, {
      method: "DELETE",
      headers: { "Content-Type": "application/json" },
    });
    printResult("TC-BACK-007", "Delete user account", deleteResponse.status, 200);
  } catch (error) {
    console.error(`${RED}[ERROR]${RESET} ${error.message}`);
  }
}

async function runTests() {
  console.log(`${BLUE}Running Backend API Tests...${RESET}`);
  await testSignup();
  await new Promise((resolve) => setTimeout(resolve, 1000));
  await testLogin();
  await new Promise((resolve) => setTimeout(resolve, 1000));
  await testAdminPermissions("user-id-placeholder"); // Replace with actual user ID
  await new Promise((resolve) => setTimeout(resolve, 1000));
  await testGetAllUsers();
  await new Promise((resolve) => setTimeout(resolve, 1000));
  await testUserManagement("johndoe");
  await new Promise((resolve) => setTimeout(resolve, 1000));
  await testUserMe();
}

runTests();

File: project/backend/test_backend_api.py
import requests
import time  # Import the time module for sleep

BASE_URL = "http://localhost:8000/api" 

# ANSI escape codes for colors
GREEN = "\033[92m"
RED = "\033[91m"
BLUE = "\033[94m"
RESET = "\033[0m"

# Global variable to store user information
USER_DICT = None

# Create a session object to persist cookies
session = requests.Session()


def print_result(test_case, description, result, expected):
    """
    Prints the test result in color-coded format using ANSI escape codes.
    """
    if result == expected:
        print(f"{GREEN}[PASS]{RESET} {test_case}: {description}")
    else:
        print(f"{RED}[FAIL]{RESET} {test_case}: {description}")
        print(f"  Expected: {expected}, Got: {result}")


# Helper function to retrieve the user dictionary
def get_user_info(username):
    global USER_DICT
    url = f"{BASE_URL}/auth/test-user-id"
    payload = {"username": username}
    response = session.post(url, json=payload)  # Use the session object
    if response.status_code == 200:
        USER_DICT = response.json().get("user")
        print(f"{BLUE}User dictionary retrieved: {USER_DICT}{RESET}")
    else:
        print(f"{RED}[FAIL]{RESET} Failed to retrieve user info. Status code: {response.status_code}")
        print(f"Response: {response.text}")


# TC-BACK-001: Verify that a new user can be created with valid data
def test_signup():
    url = f"{BASE_URL}/auth/signup"
    payload = {
        "fname": "John",
        "lname": "Doe",
        "email": "john.doe@example.com",
        "username": "johndoe",
        "password": "Password123"
    }
    response = session.post(url, json=payload)  # Use the session object
    print_result("TC-BACK-001", "Verify user signup", response.status_code, 201)


# TC-BACK-002: Test login with valid and invalid credentials
def test_login():
    url = f"{BASE_URL}/auth/login"

    # Valid credentials
    valid_payload = {"username": "johndoe", "password": "Password123"}
    valid_response = session.post(url, json=valid_payload)  # Use the session object
    print_result("TC-BACK-002", "Login with valid credentials", valid_response.status_code, 200)

    # Debugging: Print cookies stored in the session
    print(f"Session cookies: {session.cookies.get_dict()}")

    # Invalid password
    invalid_password_payload = {"username": "johndoe", "password": "WrongPassword"}
    invalid_password_response = session.post(url, json=invalid_password_payload)  # Use the session object
    print_result("TC-BACK-002", "Login with invalid password", invalid_password_response.status_code, 401)

    # User not found
    invalid_user_payload = {"username": "nonexistentuser", "password": "Password123"}
    invalid_user_response = session.post(url, json=invalid_user_payload)  # Use the session object
    print_result("TC-BACK-002", "Login with nonexistent user", invalid_user_response.status_code, 404)



def make_admin():
    if not USER_DICT:
        print(f"{RED}[FAIL]{RESET} TC-BACK-004: User dictionary not found. Run test_signup and get_user_info first.")
        return

    url = f"{BASE_URL}/admin/test-grant-admin/{USER_DICT['_id']}"
    payload = {"user": USER_DICT}  # Include the user parameter in the request payload
    response = session.post(url, json=payload)  # Use the session object
    print_result("TESTCALL", "Grant admin privileges", response.status_code, 200)

# TC-BACK-004: Test granting admin privileges
def test_grant_admin():
    """
    Test granting admin privileges to a user using session tokens.
    """
    if not USER_DICT:
        print(f"{RED}[FAIL]{RESET} TC-BACK-004: User dictionary not found. Run test_signup and get_user_info first.")
        return

    url = f"{BASE_URL}/admin/permissions/{USER_DICT['_id']}"
    response = session.post(url, headers={"Content-Type": "application/json"})  # Use the session object
    print_result("TC-BACK-004", "Grant admin privileges", response.status_code, 200)



# Add sleep between test functions
if __name__ == "__main__":
    print(f"{BLUE}Running Backend API Tests...{RESET}")
    test_signup()
    time.sleep(1)  # Add a 1-second delay
    get_user_info("johndoe")  # Retrieve the user dictionary for the test user
    time.sleep(1)  # Add a 1-second delay
    test_login()
    time.sleep(1)  # Add a 1-second delay
    test_grant_admin()
